File: app/src/main/java/com/bartovapps/gpstriprec/kmlhleper/KmlManager.java
//package com.bartovapps.gpstriprec.kmlhleper;
//
//
//import static core.trip_manager.TripManagerImpl.TRIPS_DIR;
//
//import android.content.Context;
//import android.util.Log;
//
//import com.bartovapps.gpstriprec.R;
//import com.bartovapps.gpstriprec.utils.Utils;
//import com.google.android.gms.maps.model.LatLng;
//
//import org.jdom2.Document;
//import org.jdom2.Element;
//import org.jdom2.JDOMException;
//import org.jdom2.Namespace;
//import org.jdom2.input.SAXBuilder;
//import org.jdom2.output.Format;
//import org.jdom2.output.XMLOutputter;
//
//import java.io.File;
//import java.io.FileWriter;
//import java.io.IOException;
//import java.io.InputStream;
//import java.util.ArrayList;
//import java.util.List;
//
//import javax.inject.Inject;
//import javax.inject.Singleton;
//
//import dagger.hilt.android.qualifiers.ApplicationContext;
//
//@Singleton
//public class com.bartovapps.gpstriprec.core.files.kml.com.bartovapps.gpstriprec.domain.files.kml.KmlManager {
//
////	private static final String LOG_TAG = "GPS RECORDER";
//	private static final String KML_NS = "http://www.opengis.net/kml/2.2";
//	// File xmlFile = new File("/sdcard/route.kml");
//
//	Context context;
//	private InputStream stream;
//	private SAXBuilder builder;
//	private Document doc;
//	private Element rootNode;
//
//	@Inject
//	public com.bartovapps.gpstriprec.core.files.kml.com.bartovapps.gpstriprec.domain.files.kml.KmlManager(@ApplicationContext  Context context) {
//		this.context = context;
//		this.stream = context.getResources().openRawResource(R.raw.trip_raw);
//		builder = new SAXBuilder();
//		// Log.i(LOG_TAG, "KML Helper created");
//	}
//
//	public void openRawDocument() {
//		try {
//			doc = (Document) builder.build(stream);
//			rootNode = doc.getRootElement();
//			// Log.i(LOG_TAG, "Raw Document Opened: root element:" +
//			// rootNode.getName());
//		} catch (JDOMException e) {
//			e.printStackTrace();
//		} catch (IOException e) {
//			e.printStackTrace();
//		}
//	}
//
//	public void updateStartPoint(String point) {
//		Element startPoint = null;
//		Element docElement = rootNode.getChild("Document",
//				Namespace.getNamespace(KML_NS));
//		try {
//
//			if (docElement != null) {
//				List<Element> placemarks = docElement.getChildren("Placemark",
//						Namespace.getNamespace(KML_NS));
//				// Log.i(LOG_TAG, "number of placemarks " + placemarks.size());
//
//				for (Element node : placemarks) {
//					// Log.i(LOG_TAG, "Placemark : " +
//					// node.getAttribute("id").getIntValue());
//					if (node.getAttribute("id").getIntValue() == 1) {
//						startPoint = node.getChild("Point",
//								Namespace.getNamespace(KML_NS));
//						break;
//					}
//				}
//			}
//		} catch (Exception e) {
//			e.getMessage();
//		}
//		@SuppressWarnings("unused")
//		Element coordinated = startPoint.getChild("coordinates",
//				Namespace.getNamespace(KML_NS)).setText(point);
//		// Log.i(LOG_TAG, "start point coordinates: " + coordinated.getValue());
//
//	}
//
//	public void updateEndPoint(String point) {
//		Element startPoint = null;
//		Element docElement = rootNode.getChild("Document",
//				Namespace.getNamespace(KML_NS));
//		try {
//
//			if (docElement != null) {
//				List<Element> placemarks = docElement.getChildren("Placemark",
//						Namespace.getNamespace(KML_NS));
//				// Log.i(LOG_TAG, "number of placemarks " + placemarks.size());
//
//				for (Element node : placemarks) {
//					// Log.i(LOG_TAG, "Placemark : " +
//					// node.getAttribute("id").getIntValue());
//					if (node.getAttribute("id").getIntValue() == 2) {
//						startPoint = node.getChild("Point",
//								Namespace.getNamespace(KML_NS));
//						break;
//					}
//				}
//			}
//		} catch (Exception e) {
//			e.getMessage();
//		}
//		@SuppressWarnings("unused")
//		Element coordinated = startPoint.getChild("coordinates",
//				Namespace.getNamespace(KML_NS)).setText(point);
//
//		// Log.i(LOG_TAG, "end point coordinates: " + coordinated.getValue());
//
//	}
//
//	private void updateRouteMarks(String route) {
//		Element LineString = null;
//		Element docElement = rootNode.getChild("Document",
//				Namespace.getNamespace(KML_NS));
//		try {
//
//			if (docElement != null) {
//				List<Element> placemarks = docElement.getChildren("Placemark",
//						Namespace.getNamespace(KML_NS));
//				// Log.i(LOG_TAG, "number of placemarks " + placemarks.size());
//
//				for (Element node : placemarks) {
//					// Log.i(LOG_TAG, "Placemark : " + node.getAttribute("id"));
//					if (node.getAttribute("id").getValue().equals("route")) {
//						LineString = node.getChild("LineString",
//								Namespace.getNamespace(KML_NS));
//						break;
//					}
//				}
//			}
//		} catch (Exception e) {
//			e.getMessage();
//		}
//
//		@SuppressWarnings("unused")
//		Element coordinated = LineString.getChild("coordinates",
//				Namespace.getNamespace(KML_NS)).setText(route);
//
//		// Log.i(LOG_TAG, "end point coordinates: " + coordinated.getValue());
//
//	}
//
//	public String updateTrip(ArrayList<String> locations) {
//		StringBuilder route = new StringBuilder();
//
//		String mapFile = null;
//
//		if (locations.size() > 1) {
//			updateStartPoint(locations.get(0));
//			updateEndPoint(locations.get(locations.size() - 1));
//			for (String mark : locations) {
//				route.append(mark + "\n");
//			}
//			updateRouteMarks(route.toString());
//			mapFile = writeFile();
//		}
//
//		closeKml();
//		return mapFile;
//	}
//
//	public String updateTripLatLng(List<LatLng> latlngs) {
//		StringBuilder route = new StringBuilder();
//		String mapFile = null;
//
//
//		if (latlngs.size() > 1) {
//			LatLng startPoint = latlngs.get(0);
//			LatLng stopPoint = latlngs.get(latlngs.size() - 1);
//
//			updateStartPoint(startPoint.longitude + "," + startPoint.latitude);
//			updateEndPoint(stopPoint.longitude + "," + stopPoint.latitude);
//			for (LatLng latlng : latlngs) {
//				route.append(latlng.longitude + "," + latlng.latitude + "\n");
//			}
//			// Log.i(LOG_TAG, "route coordinates: ");
//			updateRouteMarks(route.toString());
//			mapFile = writeFile();
//		}
//		return mapFile;
//	}
//
//	private String writeFile() {
//
//		long timestamp = System.currentTimeMillis();
//
//		if (Utils.checkExternalStorageState()) {
//			File fileDir = new File(context.getExternalFilesDir(null).getPath() + TRIPS_DIR );
//
//			if (!fileDir.exists()) {
//				boolean mkdirs = fileDir.mkdirs();
//				Log.i("com.bartovapps.gpstriprec.core.files.kml.com.bartovapps.gpstriprec.domain.files.kml.KmlManager", "fileDir make dir result: " + mkdirs);
//			}
//
//			String fileName = "/trip_" + timestamp + ".kml";
//
//			File kmlFile = new File(fileDir, fileName);
//
//			XMLOutputter xmlOutput = new XMLOutputter();
//			xmlOutput.setFormat(Format.getPrettyFormat());
//
//			try {
//				FileWriter fr = new FileWriter(kmlFile);
//				xmlOutput.output(doc, fr);
//				fr.flush();
//				fr.close();
//			} catch (IOException e) {
//				e.printStackTrace();
//				return "Can't save trip file";
//			}
//
//
//			return kmlFile.toString();
//		} else {
//			return "Can't save trip file";
//		}
//	}
//
//	public void closeKml(){
//		try {
//			stream.close();
//		} catch (IOException e) {
//			e.printStackTrace();
//		}
//	}
//
//}

File: app/src/main/kotlin/core/files/kml/KmlManager.kt
package com.bartovapps.gpstriprec.core.files.kml

import android.content.res.Resources
import com.bartovapps.gpstriprec.R
import com.bartovapps.gpstriprec.core.files.path_provider.PathProvider
import com.google.android.gms.maps.model.LatLng
import org.jdom2.Document
import org.jdom2.Element
import org.jdom2.JDOMException
import org.jdom2.Namespace
import org.jdom2.input.SAXBuilder
import org.jdom2.output.Format
import org.jdom2.output.XMLOutputter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class KmlManager @Inject constructor(// File xmlFile = new File("/sdcard/route.kml");
    private val resources: Resources,
    private val pathProvider: PathProvider,
    private val saxBuilder: SAXBuilder
) {
    private var doc: Document? = null
    private var rootNode: Element? = null

    fun openRawDocument() {
        try {
            resources.openRawResource(R.raw.trip_raw).use {
                doc = saxBuilder.build(it) as Document
                rootNode = doc?.rootElement
            }
        } catch (e: JDOMException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun updateStartPoint(point: String?) {
        var startPoint: Element? = null
        val docElement = rootNode?.getChild(
            "Document",
            Namespace.getNamespace(KML_NS)
        )
        try {
            if (docElement != null) {
                val placemarks = docElement.getChildren(
                    "Placemark",
                    Namespace.getNamespace(KML_NS)
                )

                for (node in placemarks) {
                    if (node.getAttribute("id").intValue == 1) {
                        startPoint = node.getChild(
                            "Point",
                            Namespace.getNamespace(KML_NS)
                        )
                        break
                    }
                }
            }
        } catch (e: Exception) {
            e.message
        }
        @Suppress("unused") val coordinated = startPoint!!.getChild(
            "coordinates",
            Namespace.getNamespace(KML_NS)
        ).setText(point)
    }

    fun updateEndPoint(point: String?) {
        var startPoint: Element? = null
        val docElement = rootNode!!.getChild(
            "Document",
            Namespace.getNamespace(KML_NS)
        )
        try {
            if (docElement != null) {
                val placemarks = docElement.getChildren(
                    "Placemark",
                    Namespace.getNamespace(KML_NS)
                )

                for (node in placemarks) {

                    if (node.getAttribute("id").intValue == 2) {
                        startPoint = node.getChild(
                            "Point",
                            Namespace.getNamespace(KML_NS)
                        )
                        break
                    }
                }
            }
        } catch (e: Exception) {
            e.message
        }
        @Suppress("unused") val coordinated = startPoint?.getChild(
            "coordinates",
            Namespace.getNamespace(KML_NS)
        )?.setText(point)
    }

    private fun updateRouteMarks(route: String) {
        var lineString: Element? = null
        val docElement = rootNode!!.getChild(
            "Document",
            Namespace.getNamespace(KML_NS)
        )
        try {
            if (docElement != null) {
                val placemarks = docElement.getChildren(
                    "Placemark",
                    Namespace.getNamespace(KML_NS)
                )

                for (node in placemarks) {
                    // Log.i(LOG_TAG, "Placemark : " + node.getAttribute("id"));
                    if (node.getAttribute("id").value == "route") {
                        lineString = node.getChild(
                            "LineString",
                            Namespace.getNamespace(KML_NS)
                        )
                        break
                    }
                }
            }
        } catch (e: Exception) {
            e.message
        }

        @Suppress("unused") val coordinated = lineString?.getChild(
            "coordinates",
            Namespace.getNamespace(KML_NS)
        )?.setText(route)

        // Log.i(LOG_TAG, "end point coordinates: " + coordinated.getValue());
    }

    fun updateTrip(locations: List<String>): String? {
        val route = StringBuilder()

        var mapFile: String? = null

        if (locations.size > 1) {
            updateStartPoint(locations[0])
            updateEndPoint(locations[locations.size - 1])
            for (mark in locations) {
                route.append(mark + "\n")
            }
            updateRouteMarks(route.toString())
            mapFile = writeFile()
        }

        return mapFile
    }

    fun updateTripLatLng(latlngs: List<LatLng>): String? {
        val route = StringBuilder()
        var mapFile: String? = null


        if (latlngs.size > 1) {
            val startPoint = latlngs[0]
            val stopPoint = latlngs[latlngs.size - 1]

            updateStartPoint(startPoint.longitude.toString() + "," + startPoint.latitude)
            updateEndPoint(stopPoint.longitude.toString() + "," + stopPoint.latitude)
            for (latlng in latlngs) {
                route.append(latlng.longitude.toString() + "," + latlng.latitude + "\n")
            }
            updateRouteMarks(route.toString())
            mapFile = writeFile()
        }
        return mapFile
    }

    private fun writeFile(): String {
        val timestamp = System.currentTimeMillis()

        val fileName = "/trip_$timestamp.kml"
        val kmlFile = File(pathProvider.provideTripKmlFilesPath(), fileName)

        val xmlOutput = XMLOutputter()
        xmlOutput.format = Format.getPrettyFormat()
        val fr = FileWriter(kmlFile)
        try {
            xmlOutput.output(doc, fr)
        } catch (e: IOException) {
            e.printStackTrace()
            return "Can't save trip file"
        }finally {
            fr.flush()
            fr.close()
        }
        return kmlFile.toString()
    }

    companion object {
        //	private static final String LOG_TAG = "GPS RECORDER";
        private const val KML_NS = "http://www.opengis.net/kml/2.2"
    }
}